samples/gibbs_ensemble/gibbs_ensemble.py

```diff
@@ -0,0 +1,154 @@
+#
+# Copyright (C) 2013-2019 The ESPResSo project
+#
+# This file is part of ESPResSo.
+#
+# ESPResSo is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ESPResSo is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Client class containing the methods used for communicating and manipulating the system
+MessageID class for communication identifiers
+"""
+
+from enum import unique, Enum
+from multiprocessing import Process
+import os
+import numpy as np
+
+
+@unique
+class MessageID(Enum):
+    # Message identifiers
+    ENERGY = 0
+    END = 1
+    MOVE_PART = 2
+    MOVE_PART_REVERT = 21
+    CHANGE_VOLUME = 3
+    CHANGE_VOLUME_REVERT = 31
+    PART_ADD = 4
+    PART_ADD_REVERT = 41
+    PART_REMOVE = 5
+    PART_REMOVE_REVERT = 51
+    INFO = 6
+
+
+class Client(Process):
+    def __init__(self,
+                 box_name,
+                 pipe,
+                 seed,
+                 box_length,
+                 num_part):
+        super().__init__(name=box_name)
+        self.init_box_length = box_length
+        self.init_num_part = num_part
+        self.system = None
+        self.old_box_length = 0.0
+        # Pipe for communicating
+        self.pipe = pipe
+        self.seed = seed
+
+    @property
+    def density(self):
+        return len(self.system.part[:].id) / float(np.prod(self.system.box_l))
+
+    @property
+    def energy(self):
+        return self.system.analysis.energy()["total"]
+
+    def send_energy(self):
+        self.pipe.send([MessageID.ENERGY, self.energy])
+
+    def end_simulation(self):
+        # Additional simulation checkpointing etc.
+        os._exit(os.EX_OK)
+
+    def send_info(self):
+        self.pipe.send([MessageID.INFO, self.system.box_l[0],
+                        len(self.system.part[:].id)])
+
+    def move_particle(self, DX_MAX):
+        """ Move random particle inside the box """
+        self.old_id = np.random.choice(self.system.part[:].id)
+        self.old_pos = self.system.part[self.old_id].pos
+        self.system.part[self.old_id].pos = self.old_pos + \
+            (0.5 - np.random.random(size=3)) * DX_MAX
+        self.send_energy()
+
+    def revert_move_particle(self):
+        """ Revert particle move """
+        self.system.part[self.old_id].pos = self.old_pos
+
+    def add_particle(self):
+        """ Add a particle at random inside box """
+        self.system.part.add(pos=np.random.random(3) * self.system.box_l)
```

Comment:
Please store the handle to the created particle in the class and use it in the revert step. Otherwise, you might remove a different particle than what you added in case of a revert

Reply:
I like using handles to store the particle data, but why not use it for the other methods as well? This way one doesn't have to store the data with the to_dict() method

Reply:
I included what I meant in 2377506, please review it again.